src/script/interpreter.cpp

```diff
@@ -1781,6 +1910,136 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
     return true;
 }
 
+template <class T>
+bool GenericTransactionSignatureChecker<T>::CheckVaultSpendToRecoveryOutputs(
+    const uint256& recovery_spk_hash) const
+{
+    if (!this->txdata) return HandleMissingData(m_mdb);
+    const auto& txd{*this->txdata};
+
+    if (!CommonVaultSpendChecks(txd, this->txTo->vout)) {
+        return false;
+    }
+    const CTxOut& value_out = this->txTo->vout[0];
+
+    return (VaultScriptHash(HASHER_VAULT_RECOVERY_SPK, value_out.scriptPubKey) ==
+            recovery_spk_hash);
+}
+
+template <class T>
+bool GenericTransactionSignatureChecker<T>::CheckUnvaultTriggerOutputs(
+    const uint256& recovery_spk_hash,
+    const CScriptNum& spend_delay,
+    bool require_minimal) const
+{
+    if (!this->txdata) return HandleMissingData(m_mdb);
+    const auto& txd{*this->txdata};
+
+    if (!CommonVaultSpendChecks(txd, this->txTo->vout)) {
+        return false;
+    }
+    const CTxOut& value_out = this->txTo->vout[0];
+    const CScript& value_spk = value_out.scriptPubKey;
+
+    auto pc = value_spk.begin();
+    valtype pushdata;
+    opcodetype opcode;
+
+    // Check for 32 bytes of recovery-spk-hash.
+    if (!value_spk.GetOp(pc, opcode, pushdata) || opcode != 0x20) {
+        return false;
+    }
+    uint256 proposed_recovery_spk_hash{pushdata};
+
+    if (proposed_recovery_spk_hash != recovery_spk_hash) {
+        return false;
+    }
+
+    // Check for spend-delay; must be between 1 and 4 bytes.
+    if (!value_spk.GetOp(pc, opcode, pushdata)) {
+        return false;
+    }
+    CScriptNum proposed_spend_delay{0};
+
+    // TODO: this is ugly, but basically we have to replicate EvalScript's parsing
+    // of numbers here. If the number is between [1, 16], it will be OP_[n], but
+    // if it's larger it's encoded as pushdata.
+    if (opcode >= OP_1 && opcode <= OP_16) {
+        proposed_spend_delay = (int)opcode - (int)(OP_1 - 1);
+    } else {
+        if (pushdata.size() < 1 || pushdata.size() > CScriptNum::nDefaultMaxNumSize) {
+            return false;
+        }
+        proposed_spend_delay = CScriptNum(pushdata, require_minimal);
+    }
+
+    if (proposed_spend_delay != spend_delay) {
+        return false;
+    }
+
+    // Check for target-hash; we don't care what it is as long as it's 32 bytes.
+    if (!value_spk.GetOp(pc, opcode) || opcode != 0x20) {
+        return false;
+    }
+
+    // Finally, check that OP_UNVAULT is the last opcode.
+    if (!value_spk.GetOp(pc, opcode) || opcode != OP_UNVAULT || pc != value_spk.end()) {
+        return false;
+    }
+
+    return true;
+}
+
+template <class T>
+bool GenericTransactionSignatureChecker<T>::CheckUnvaultTarget(
+    const uint256& target_outputs_hash) const
+{
+    // We can't use precomputed transaction data here because, since the input lacks a
+    // witness, the precomputation routines don't run. I.e. `txdata.hashOutputs` is blank.
+    return SHA256Uint256(GetOutputsSHA256(*this->txTo)) == target_outputs_hash;
```

Comment:
In fbf344931a2224dc3ace0011c1a4cb5cd4187f3e "wip: draft implementation"

Note that this is a different serialization from the one described in the paper. The paper says to serialize the scriptPubKey first, then nValue. However `GetOutputsSHA256` serializes each output as nValue then scriptPubKey. I'm going to assume this is an error in the paper as I think it is better the way it is implemented rather than described.